File: src/main/java/cn/academy/vanilla/electromaster/skill/MagManip.java
/**
 * Copyright (c) Lambda Innovation, 2013-2015
 * 本作品版权由Lambda Innovation所有。
 * http://www.li-dev.cn/
 *
 * This project is open-source, and it is distributed under  
 * the terms of GNU General Public License. You can modify
 * and distribute freely as long as you follow the license.
 * 本项目是一个开源项目，且遵循GNU通用公共授权协议。
 * 在遵照该协议的情况下，您可以自由传播和修改。
 * http://www.gnu.org/licenses/gpl.html
 */
package cn.academy.vanilla.electromaster.skill;

import cn.academy.ability.api.Skill;
import cn.academy.ability.api.ctrl.ActionManager;
import cn.academy.ability.api.ctrl.Cooldown;
import cn.academy.ability.api.ctrl.SkillInstance;
import cn.academy.ability.api.ctrl.action.SkillSyncAction;
import cn.academy.ability.api.data.AbilityData;
import cn.academy.core.entity.EntityBlock;
import cn.academy.vanilla.electromaster.CatElectromaster;
import cn.academy.vanilla.electromaster.entity.EntitySurroundArc;
import cn.academy.vanilla.electromaster.entity.EntitySurroundArc.ArcType;
import cn.lambdalib.annoreg.core.Registrant;
import cn.lambdalib.annoreg.mc.RegEntity;
import cn.lambdalib.util.entityx.event.CollideEvent;
import cn.lambdalib.util.entityx.event.CollideEvent.CollideHandler;
import cn.lambdalib.util.entityx.handlers.Rigidbody;
import cn.lambdalib.util.helper.EntitySyncer;
import cn.lambdalib.util.helper.Motion3D;
import cn.lambdalib.util.helper.EntitySyncer.Synchronized;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.block.Block;
import net.minecraft.command.IEntitySelector;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.DamageSource;
import net.minecraft.world.World;

/**
 * Magnet manipulation
 * @author WeAthFolD
 */
@Registrant
public class MagManip extends Skill {

	public static final MagManip instance = new MagManip();
	
	private MagManip() {
		super("mag_manip", 2);
	}
	
	static boolean accepts(AbilityData data, ItemStack stack) {
		Block block; 
		if(stack == null || (block = Block.getBlockFromItem(stack.getItem())) == null)
			return false;
		
		// Originally I want to strenghten this limit, but this skill would be too useless then.
		// So currently, just blocks with rock | anvil material can be used.
		return CatElectromaster.isWeakMetalBlock(block);
	}
	
	static float getDamage(AbilityData data) {
		return instance.callFloatWithExp("damage", data);
	}
	
	static float getExpIncr() {
		return instance.getFloat("expincr");
	}
	
	@Override
	public SkillInstance createSkillInstance(EntityPlayer player) {
		return new SkillInstance().addChild(new ManipAction());
	}
	
	public static class ManipAction extends SkillSyncAction {

		public ManipAction() {
			super(-1);
		}
		
		@Override
		public void onStart() {
			super.onStart();
			
			if(!isRemote && !checkItem()) {
				ActionManager.abortAction(this);
			}
			
			if(isRemote)
				startEffect();
		}
		
		@Override
		public void onTick() {
			if(!isRemote) {
				if(!checkItem() || !cpData.canPerform(instance.getConsumption(aData))) {
					ActionManager.abortAction(this);
				}
			} else
				updateEffect();
		}
		
		@Override
		public void onEnd() {
			if(!isRemote) {
				if(checkItem()) {
					cpData.performWithForce(instance.getOverload(aData), instance.getConsumption(aData));
					
					ItemStack stack = player.getCurrentEquippedItem();
					EntityBlock entity = new ManipEntityBlock(player, getDamage(aData));
					
					entity.fromItemStack(stack);
					new Motion3D(player, true).multiplyMotionBy(1.6).applyToEntity(entity);
					
					if(entity.isAvailable()) {
						if(!player.capabilities.isCreativeMode) {
							if(--stack.stackSize == 0) {
								player.setCurrentItemOrArmor(0, null);
							}
						}
						player.worldObj.spawnEntityInWorld(entity);
					}
					
					aData.addSkillExp(instance, getExpIncr());
				}
			}
			
			setCooldown(instance, instance.getCooldown(aData));
		}
		
		@Override
		public void onFinalize() {
			if(isRemote)
				endEffect();
		}
		
		private boolean checkItem() {
			return accepts(aData, player.getCurrentEquippedItem());
		}
		
		// CLIENT
		@SideOnly(Side.CLIENT)
		EntitySurroundArc arc;
		
		@SideOnly(Side.CLIENT)
		private void startEffect() {
			arc = new EntitySurroundArc(player);
			arc.setArcType(ArcType.NORMAL);
			arc.life = 233333;
			player.worldObj.spawnEntityInWorld(arc);
			player.capabilities.setPlayerWalkSpeed(0.05f);
		}
		
		@SideOnly(Side.CLIENT)
		private void updateEffect() {
		}
		
		@SideOnly(Side.CLIENT)
		private void endEffect() {
			player.capabilities.setPlayerWalkSpeed(0.1f);
			arc.setDead();
		}
		
	}
	
	
	@RegEntity
	public static class ManipEntityBlock extends EntityBlock {
		
		EntitySyncer syncer;
		
		@Synchronized
		EntityPlayer player;
		
		float damage;
		
		public ManipEntityBlock(EntityPlayer _player, float _damage) {
			super(_player);
			player = _player;
			damage = _damage;
		}

		public ManipEntityBlock(World world) {
			super(world);
		}
		
		@Override
		public void onFirstUpdate() {
			super.onFirstUpdate();
			(syncer = new EntitySyncer(this)).init();
			
			Rigidbody rb = this.getMotionHandler(Rigidbody.class);
			rb.entitySel = new IEntitySelector() {
				@Override
				public boolean isEntityApplicable(Entity target) {
					return target != player;
				}
			};
			rb.gravity = 0.05;
			
			this.regEventHandler(new CollideHandler() {

				@Override
				public void onEvent(CollideEvent event) {
					if(!worldObj.isRemote && 
						event.result != null && 
						event.result.entityHit != null)
						event.result.entityHit.attackEntityFrom(DamageSource.causePlayerDamage(player), damage);
				}
				
			});
			
			if(worldObj.isRemote)
				startClient();
		}
		
		@SideOnly(Side.CLIENT)
		private void startClient() {
			EntitySurroundArc surrounder = new EntitySurroundArc(this);
			surrounder.life = 30;
			surrounder.setArcType(ArcType.THIN);
			worldObj.spawnEntityInWorld(surrounder);
		}
		
		@Override
		public void onUpdate() {
			super.onUpdate();
			syncer.update();
		}
		
	}

}

File: src/main/scala/cn/academy/vanilla/electromaster/skill/MagManip.scala
package cn.academy.vanilla.electromaster.skill

import java.util.function.Predicate

import cn.academy.ability.api.context._
import cn.academy.ability.api.{AbilityContext, Skill}
import cn.academy.core.client.sound.{ACSounds, FollowEntitySound}
import cn.academy.core.entity.EntityBlock
import cn.academy.vanilla.electromaster.CatElectromaster
import cn.academy.vanilla.electromaster.entity.EntitySurroundArc
import cn.academy.vanilla.electromaster.entity.EntitySurroundArc.ArcType
import cn.lambdalib.annoreg.core.Registrant
import cn.lambdalib.annoreg.mc.RegEntity
import cn.lambdalib.multiblock.BlockMulti
import cn.lambdalib.s11n.network.NetworkMessage.Listener
import cn.lambdalib.util.entityx.MotionHandler
import cn.lambdalib.util.entityx.event.CollideEvent
import cn.lambdalib.util.entityx.event.CollideEvent.CollideHandler
import cn.lambdalib.util.entityx.handlers.Rigidbody
import cn.lambdalib.util.generic.{MathUtils, RandUtils}
import cn.lambdalib.util.helper.EntitySyncer
import cn.lambdalib.util.helper.EntitySyncer.Synchronized
import cn.lambdalib.util.mc._
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.block.{BlockDoor, Block}
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.world.World

private[electromaster] object MagManip extends Skill("mag_manip", 2) {

  @SideOnly(Side.CLIENT)
  override def activate(rt: ClientRuntime, keyid: Int) = activateSingleKey(rt, keyid, p => new MagManipContext(p))

  private[electromaster] def accepts(player: EntityPlayer, block: Block) = block match {
    case _: BlockMulti => false // Avoid jerky result for multiblock structure.
    case _: BlockDoor => false
    case _ => CatElectromaster.isMetalBlock(block)
  }

}

private object MagManipContext {
  val StateMoving = 0
  val StateCharging = 1

  final val MSG_PERFORM = "perform"
}

import cn.academy.ability.api.AbilityAPIExt._
import cn.academy.vanilla.electromaster.skill.MagManip._
import cn.academy.vanilla.electromaster.skill.MagManipContext._
import cn.lambdalib.util.mc.MCExtender._

private class MagManipContext(p: EntityPlayer) extends Context(p, MagManip) with IConsumptionProvider {

  private val consumption = MathUtils.lerpf(225, 275, ctx.getSkillExp)
  private val overload = MathUtils.lerpf(72, 33, ctx.getSkillExp)
  private val cooldown = MathUtils.lerpf(60, 40, ctx.getSkillExp).toInt
  private val damage = MathUtils.lerpf(18, 30, ctx.getSkillExp)
  private val speed = MathUtils.lerpf(0.5f, 1.0f, ctx.getSkillExp)

  var state = StateMoving

  var entity: MagManipEntityBlock = _

  var performed = false

  @Listener(channel=MSG_KEYUP, side=Array(Side.CLIENT))
  def l_keyUp() = {
    sendToServer(MSG_PERFORM)
  }

  @Listener(channel=MSG_KEYABORT, side=Array(Side.CLIENT))
  def l_keyAbort() = {
    terminate()
  }

  @Listener(channel=MSG_MADEALIVE, side=Array(Side.SERVER))
  def s_makeAlive() = {
    // Find the block to use.
    val stack = player.getCurrentEquippedItem
    Option(stack)
      .flatMap(s => Option(Block.getBlockFromItem(s.getItem))) match {
      case Some(block) if accepts(player, block) =>
        if (!player.capabilities.isCreativeMode) {
          stack.stackSize -= 1
          if (stack.stackSize == 0) {
            player.setCurrentItemOrArmor(0, null)
          }
        }

        entity = new MagManipEntityBlock(player, 10)
        entity.setBlock(block)
        entity.setPos(player.headPosition)

        world.spawnEntityInWorld(entity)

        updateMoveTo()

      case _ =>
        val trace: TraceResult = Raytrace.traceLiving(player, 10, EntitySelectors.nothing(), new IBlockSelector {
          override def accepts(world: World, x: Int, y: Int, z: Int, block: Block): Boolean = {
            MagManip.accepts(player, block)
          }
        })

        trace match {
          case res: BlockResult =>
            val block = res.getBlock(world)

            val (x, y, z) = res.pos
            world.setBlockToAir(x, y, z)

            entity = new MagManipEntityBlock(player, 10)
            entity.setBlock(block)
            entity.setPosition(x + .5, y + .5, z + .5)

            world.spawnEntityInWorld(entity)

            updateMoveTo()
          case _ => terminate()
        }
    }
  }

  @Listener(channel=MSG_TICK, side=Array(Side.SERVER))
  def s_tick() = {
    updateMoveTo()
  }

  @Listener(channel=MSG_PERFORM, side=Array(Side.SERVER))
  def s_perform() = {
    performed = true
    entity.actionType = entity.ActNothing
    entity.setPlaceFromServer(true)

    val distsq = player.getDistanceSqToEntity(entity)
    if (distsq < 25 && ctx.consume(overload, consumption)) {
      val pos = Raytrace.getLookingPos(player, 20).getLeft
      val delta = pos - entity.position
      entity.setVel(delta.normalize() * speed)

      ctx.setCooldown(cooldown)
      ctx.addSkillExp(0.005f)

      sendToClient(MSG_PERFORM)
    }

    terminate()
  }

  @Listener(channel=MSG_TERMINATED, side=Array(Side.SERVER))
  def s_terminate() = {
    if (!performed && entity != null) {
      entity.actionType = entity.ActNothing
      entity.setPlaceFromServer(true)
    }
  }

  private def updateMoveTo() = {
    val origin = player.headPosition - Vec3(0, 0.1, 0)
    val look = player.lookVector

    var look2 = Vec3(look)
    look2.yCoord = 0
    look2 = look2.normalize()
    look2.rotateAroundY(90)

    val pos = origin + player.lookVector * 2.0

    entity.setMoveTo(pos.x, pos.y, pos.z)
  }

  private def wrap(args: Any*) = args.map(_.asInstanceOf[AnyRef])

  override def getConsumptionHint: Float = consumption
}

@Registrant
@SideOnly(Side.CLIENT)
@RegClientContext(classOf[MagManipContext])
class MagManipContextC(par: MagManipContext) extends ClientContext(par) {

  val loopSound = new FollowEntitySound(par.player, "em.lf_loop").setLoop()

  @Listener(channel=MSG_MADEALIVE, side=Array(Side.CLIENT))
  def c_makeAlive() = {
    ACSounds.playClient(loopSound)
  }

  @Listener(channel=MSG_TERMINATED, side=Array(Side.CLIENT))
  def c_terminate() = {
    loopSound.stop()
  }

  @Listener(channel=MSG_PERFORM, side=Array(Side.CLIENT))
  def c_perform() = {
    ACSounds.playClient(player, "em.mag_manip", 1.0f)
  }


}

@Registrant
@RegEntity(freq = 10)
class MagManipEntityBlock(world: World) extends EntityBlock(world) {

  val ActNothing = 0
  val ActMoveTo  = 1

  var syncer: EntitySyncer = _

  var damage: Float = _

  @Synchronized
  var player2: EntityPlayer = null

  val yawSpeed: Float = RandUtils.rangef(1, 3)
  val pitchSpeed: Float = RandUtils.rangef(1, 3)

  @Synchronized
  var actionType: Int = _

  @Synchronized var tx: Float = _
  @Synchronized var ty: Float = _
  @Synchronized var tz: Float = _

  placeWhenCollide = false

  def this(player: EntityPlayer, damage: Float) = { this(player.worldObj)
    constructServer(player)
    this.damage = damage
    this.player2 = player
  }

  def setMoveTo(x: Double, y: Double, z: Double) = {
    actionType = ActMoveTo
    tx = x.toFloat
    ty = y.toFloat
    tz = z.toFloat
  }

  def stopMoveTo() = {
    actionType = ActNothing
  }

  override def entityInit() = {
    super.entityInit()
    syncer =new EntitySyncer(this)
    syncer.init()
  }

  override def onFirstUpdate() = {
    super.onFirstUpdate()

    val rb = getMotionHandler(classOf[Rigidbody])
    rb.entitySel = new Predicate[Entity] {
      override def test(t: Entity): Boolean = t != player2
    }

    regEventHandler(new CollideHandler {
      override def onEvent(event: CollideEvent): Unit = {
        if (!worldObj.isRemote && event.result != null && event.result.entityHit != null) {
          AbilityContext.of(player2, MagManip).attack(event.result.entityHit, damage)
        }
      }
    })

    if (worldObj.isRemote) {
      startClient()
    }
  }

  override def onUpdate() = {
    syncer.update()

    super.onUpdate()

    yaw += yawSpeed
    pitch += pitchSpeed

    actionType match {
      case ActMoveTo =>
        val dist = this.getDistanceSq(tx, ty, tz)
        val delta = Vec3(tx - posX, ty - posY, tz - posZ).normalize()
        val mo = delta * 0.2 * (dist match {
          case d if d < 4 => d / 4
          case _ => 1.0
        })

        this.setVel(mo)

      case ActNothing =>
        motionY -= 0.04
    }

    posX += motionX
    posY += motionY
    posZ += motionZ
  }

  private def move(fr: Double, to: Double, max: Double) = {
    val delta = to - fr
    fr + math.min(math.abs(delta), max) * math.signum(delta)
  }

  @SideOnly(Side.CLIENT)
  private def startClient() = {
    val surrounder = new EntitySurroundArc(this)

    surrounder.life = 233333
    surrounder.addMotionHandler(new MotionHandler[Entity] {
      override def getID: String = "killer"
      override def onUpdate(): Unit = if (MagManipEntityBlock.this.isDead) getTarget.setDead()
      override def onStart(): Unit = {}
    })
    surrounder.setArcType(ArcType.THIN)
    worldObj.spawnEntityInWorld(surrounder)
  }
}